Srcs/heiftojpeg/heiftojpeg.cpp

```diff
@@ -0,0 +1,227 @@
+    }
+
+    ImageFileReaderInterface::IdVector exifItemIds;
+    ImageFileReaderInterface::DataVector exifData;
+    reader.getReferencedToItemListByType(contextId, gridItemId, "cdsc", exifItemIds);
+    if (VERBOSE) {
+        cout << "found " << exifItemIds.size() << " cdsc items\n";
+    }
+    reader.getItemData(contextId, exifItemIds.at(0), exifData);
+    if (VERBOSE) {
+        cout << "read " << exifData.size() << " bytes of exif data\n";
+    }
+
+    ImageFileReaderInterface::IdVector tileItemIds;
+    reader.getItemListByType(contextId, "master", tileItemIds);
+    if (VERBOSE) {
+        cout << "found " << tileItemIds.size() << " tile images\n";
+    }
+
+    // Always reuse the parameter set from the first tile, sometimes tile 7 or 8 is corrupted
+    HevcImageFileReader::ParameterSetMap parameterSet;
+    reader.getDecoderParameterSets(contextId, tileItemIds.at(0), parameterSet);
+    string codeType = reader.getDecoderCodeType(contextId, tileItemIds.at(0));
+    ImageFileReaderInterface::DataVector parametersData;
+    if ((codeType == "hvc1") || (codeType == "lhv1")) {
+        // VPS (HEVC specific)
+        parametersData.insert(parametersData.end(), parameterSet.at("VPS").begin(), parameterSet.at("VPS").end());
+    }
+
+    if ((codeType == "avc1") || (codeType == "hvc1") || (codeType == "lhv1")) {
+        // SPS and PPS
+        parametersData.insert(parametersData.end(), parameterSet.at("SPS").begin(), parameterSet.at("SPS").end());
+        parametersData.insert(parametersData.end(), parameterSet.at("PPS").begin(), parameterSet.at("PPS").end());
+    } else {
+        // No other code types supported
+        throw ImageFileReaderInterface::FileReaderException(ImageFileReaderInterface::FileReaderException::StatusCode::UNSUPPORTED_CODE_TYPE);
+    }
+
+    ImageFileReaderInterface::DataVector itemDataWithDecoderParameters;
+    ImageFileReaderInterface::DataVector itemData;
+
+    vector<Magick::Image> tileImages;
+    for (auto& tileItemId: tileItemIds) {
+        itemDataWithDecoderParameters.clear();
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), parametersData.begin(), parametersData.end());
+
+        itemData.clear();
+        reader.getItemData(contextId, tileItemId, itemData);
+
+        // +1 comes from skipping first zero after decoder parameters
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), itemData.begin() + 1, itemData.end());
+
+        Magick::Image image;
+        decodeData(itemDataWithDecoderParameters, &image);
+        tileImages.push_back(image);
+    }
+
+    Magick::Montage montageOptions;
+    montageOptions.tile("8x6");
+    montageOptions.geometry("512x512");
+    list<Magick::Image> montage;
+    Magick::montageImages(&montage, tileImages.begin(), tileImages.end(), montageOptions);
+    Magick::Image image = montage.front();
+    image.magick("JPEG");
```

Comment:
before we let this run wild you should sanity check that it's default of detecting a good quality compression for our source image is working so it's not doing any chroma downsampling https://www.imagemagick.org/script/command-line-options.php#quality  maybe you can run it on the command line with -verbose and get it to tell you?

Reply:
`-verbose` didn't provide extra info, but I was able to determine based on empirical experiment that ImageMagick is using `-quality 92`, which is the default if it cannot be guessed from the file. I made this option explicit in the code. No chroma downsampling should occur for any value above 90.